src/Data/IxSet/Typed.hs

```diff
@@ -847,36 +884,60 @@ innerJoinUsing (IxSet _ ixs1) (IxSet _ ixs2) _ = f (access ixs1) (access ixs2)
 -- Lenses and other optics
 --------------------------------------------------------------------------
 
--- | Assuming the given index is a /primary key/, constructs a lens to access
--- the value associated with the primary key. The getting operation uses 'getEQ'
--- and 'getOne' and the setting operation uses 'updateIx' or 'deleteIx'.
--- Therefore, this /will/ violate lens laws if the given index is not actually a
--- primary key.
-atPrimaryKey :: (Indexable ixs a, IsIndexOf ix ixs, Functor f) => ix -> (Maybe a -> f (Maybe a)) -> (IxSet ixs a -> f (IxSet ixs a))
-atPrimaryKey i afb s' = sbt s' <$> afb (sa s')
+atPrimaryKey' :: (Indexable ixs a, IsIndexOf ix ixs, Functor f) =>
+  (IxSet ixs a -> Maybe a) -> ix -> (Maybe a -> f (Maybe a)) -> (IxSet ixs a -> f (IxSet ixs a))
+atPrimaryKey' get i afb s' = sbt s' <$> afb (sa s')
   where
-    sa = getOne . getEQ i
+    sa = get . getEQ i
     {-# INLINE sa #-}
 
     sbt s Nothing = deleteIx i s
     sbt s (Just b) = updateIx i b s
     {-# INLINE sbt #-}
-{-# INLINE atPrimaryKey #-}
+{-# INLINE atPrimaryKey' #-}
 
--- | Assuming the given index is a /primary key/, constructs a traversal to
--- access the value associated with the primary key. This will not work when the
--- provided index is not actually a primary key.
-ixPrimaryKey :: (Indexable ixs a, IsIndexOf ix ixs, Applicative f) => ix -> (a -> f a) -> (IxSet ixs a -> f (IxSet ixs a))
-ixPrimaryKey i afb =
-  atPrimaryKey
+ixPrimaryKey' :: (Indexable ixs a, IsIndexOf ix ixs, Applicative f) =>
+  (IxSet ixs a -> Maybe a) -> ix -> (a -> f a) -> (IxSet ixs a -> f (IxSet ixs a))
+ixPrimaryKey' get i afb =
+  atPrimaryKey' get
     i
     ((\x ->
         case x of
           Nothing -> pure Nothing
           Just y -> Just <$> y) .
      fmap afb)
+{-# INLINE ixPrimaryKey' #-}
+
+-- | Assuming the given index is a /primary key/, constructs a traversal to
```

Comment:
This docstring is now misleading. Primary keys are supposed to be unique.

Reply:
In fact, why not just change the `atPrimaryKey` and `ixPrimaryKey` to use the new `getUnique` dingus and not add `ixUniqueKey`?

Reply:
I wouldn't mind doing that, but there's a bunch of code which currently hums away, possibly silently ignoring duplicates if they happen. If I switch over, they'll be throwing errors now. Kind of scary.

Reply:
Another option is to pull this scary feature into local IxSet utils, and keep the library API simple.

Reply:
> Another option is to pull this scary feature into local IxSet utils, and keep the library API simple.

This.